crypto: add Ed25519->X25519 conversion functions

jeffro256 opened the pull request:
Part of upstreaming Carrot

Depends on #10111

jeffro256 committed:
crypto: add Ed25519->X25519 conversion functions

Reviewed-by: koe <ukoe@protonmail.com>
Reviewed-by: j-berman <justinberman@protonmail.com>

jeffro256 commented:
Removed redundant code by depending on `fe_frombytes_vartime()` in #10111.

vtnerd commented on src/crypto/crypto-ops.c:
```diff
@@ -3829,6 +3835,41 @@ int sc_isnonzero(const unsigned char *s) {
+  fe Z;
+  fe_1(Z);
+
+  edwardsYZ_to_x25519(xbytes, Y, Z);
```
So we "lose" a bit from the x-coordinate then? I recall this being how it worked last I checked. I guess it doesn't matter as all of the x25519 implementations ignore it anyway.

jeffro256 replied:
Yes, exactly. X25519 (and other Montgomery ladder curve implementations) choose to lose 1 bit of security in favor of the much faster point arithmetic.

jeffro256 commented:
Rebased (to fix unrelated compilation issues), and implemented feedback by Cypherstack in section 5 of https://github.com/cypherstack/fcmp-impl-audit/blob/main/FCMP_Code_Implementation_Audit.pdf.

jeffro256 commented:
Rebased and squashed

UkoeHB commented:
What paper is this conversion from?

jeffro256 commented:
@UkoeHB added a link to RFC 7748

UkoeHB commented on src/crypto/crypto-ops.c:
```diff
@@ -3904,6 +3905,39 @@ int sc_isnonzero(const unsigned char *s) {
+  int r;
+  fe_add(tmp0, Z, Y);       // Z + Y
+  fe_sub(tmp1, Z, Y);       // Z - Y
+  r = -!fe_isnonzero(tmp1); // 0 ?= (Z - Y) AKA is identity point?
```
This is only identity if `X == 0` https://eprint.iacr.org/2008/522.pdf sections 2-3

jeffro256 replied:
Non-canonical encodings of the identity point (-0) should also be rejected, right? At any rate, I can't invert `tmp1`, regardless of what $X$ is.

UkoeHB replied:
Would `-0` be a 'valid field element representation' according to the precondition (i.e. doesn't that imply reduced)?

Just objecting to the comment, not the logic. Unless it's the case that `Z == Y` is only true for identity among all valid points, which isn't immediately clear to me.

jeffro256 replied:
Ah I see, yeah I can update the comment

jeffro256 replied:
For all $Y,Z \neq 0$, $y = Y / Z = 1$, which would make it the identity point. $Z = 0, Y = 0$ is not a valid Ed25519 repr, but can happen in the code.

jeffro256 replied:
Okay done in latest

UkoeHB replied:
> For all Y , Z ≠ 0 , y = Y / Z = 1 , which would make it the identity point.

Yeah still uncertain here. Is there a paper reference or algebraic argument that you can't have a valid non-identity point with `X != 0` and `y = Y / Z = 1`?

UkoeHB replied:
I guess you could say it *must* be so because otherwise the birational map wouldn't work. Though that doesn't prove it directly.

Looking at the ed5519 curve, setting `y = 1`
```
ax^2 + y^2 = 1 + dx^2y^2
ax^2 + 1 = 1 + dx^2
ax^2 = dx^2
ax^2 - dx^2 = 0
x = 0
```
So `x` is always zero when `y = 1`.

jeffro256 replied:
Yes, exactly. We know that $Z = 0$ is an invalid representation by definition. All other cases lead to $y = 1$. Then subbing $y = 1$ gives you the chain of simplications you listed out. We just need to additionally make sure that $(a - d) \neq 0$, which is the case for Ed25519.

UkoeHB commented on src/crypto/crypto-ops.h:
```diff
@@ -165,6 +162,30 @@ void sc_muladd(unsigned char *s, const unsigned char *a, const unsigned char *b,
+ * @param h Ed25519 point in projective representation
+ * @return 0 on success, otherwise non-0 on failure
+ *
+ * Returns failure on Ed25519's identity point. Assumes `h` has valid field element representations,
+ * that Z != 0, and that Y/Z is a valid y-coordinate for Ed25519. Returns success iff Y != Z, i.e.
+ * not an identity point. The runtime is constant.
+ */
```
```suggestion
 * Returns failure on Ed25519's identity point. Assumes `h` has valid field element representations,
 * that Z != 0, and that Y/Z is a valid y-coordinate for Ed25519. Returns success iff Y != Z (Y == Z
 * corresponds with the identity point in valid point representations). The runtime is constant.
 */
```
This should fix the ambiguity that's bugging me.
